pkg/fieldpath/paved.go

```diff
@@ -220,7 +220,15 @@ func (p *Paved) GetBool(path string) (bool, error) {
 	return b, nil
 }
 
+// NOTE(muvaf): If there is no CRD, unstructured.Unstructured reads numbers as
+// float64. However, in practice, use of float64 is discouraged and when you fetch
+// an instance of a CRD whose number fields are int64, you'll get int64. So,
+// it's not really possible to test this without an api-server but that's the
+// actual behavior.
+
 // GetNumber value of the supplied field path.
+// Deprecated: Use of float64 is discouraged. Please use GetInteger.
```

Comment:
While Kubernetes API conventions discourage float64, I'm not sure we can get away with deprecating this because I think the value will actually contain float64 sometimes. When we call `SetValue` [we round trip the value through JSON encoding](https://github.com/crossplane/crossplane-runtime/blob/9b321c/pkg/fieldpath/paved.go#L238) which will make it a `float64`. The same is true when the DefaultUnstructuredConvert is running with mismatchDetection enabled.

Reply:
It seems like the number stuff is the only reason apimachinery uses a custom json unmarshaller. Maybe we should switch to that? Also, do you think that round trip is really needed? I couldn't immediately see a case where the given map is not a valid one and user doesn't get an error if we don't do the round-trip.

Reply:
Yeah, let's definitely switch this to the custom JSON unmarshaller that apimachinery is using.

We need this round trip for at least the following situation:

```go
type Cool struct {
  Coolness int
}

// We need &Cool{} to be converted to a map[string]interface{}
// or else GetValue won't be able to return its fields.
p.SetValue("very.cool.field", &Cool{})
```

Reply:
Given that `float64` is discouraged and the JSON marshaller tries hard to use `int64` instead I feel okay deprecating this function. It seems like it's still _possible_ to use a `float64` (against conventions), but we could use `GetValue` if we _really_ wanted a `float64`.

Reply:
Heads up, I'm happy to approve and merge this once the JSON unmarshaller we're using in `setValue` is changed to use the Kubernetes one.